src/main/scala/zio/telemetry/Telemetry.scala

```diff
@@ -0,0 +1,213 @@
+      r <- span(zio, child, tagError, logError)
+    } yield r
+
+  def span[R, R1 <: R with COT, E, A](
+      zio: ZIO[R, E, A],
+      opName: String,
+      tagError: Boolean = true,
+      logError: Boolean = true
+  ): ZIO[R1, E, A] =
+    for {
+      tracer <- getTracer
+      old <- getSpan
+      child <- UIO(tracer.buildSpan(opName).asChildOf(old).start())
+      r <- span(zio, child, tagError, logError)
+    } yield r
+
+  def span[R, R1 <: R with COT, E, A](
+      zio: ZIO[R, E, A],
+      span: Span,
+      tagError: Boolean,
+      logError: Boolean
+  ): ZIO[R1, E, A] =
+    ZManaged
+      .make[R1, E, Span](getSpan <* setSpan(span)) { old =>
+        getCurrentTimeMicros.flatMap(now => UIO(span.finish(now))) *> setSpan(
+          old
+        )
+      }
+      .use(
+        _ =>
+          zio.catchAllCause { cause =>
+            tag("error", true).when(tagError) *>
+              log(
+                Map("error.object" -> cause, "stack" -> cause.prettyPrint)
+              ).when(logError) *>
+              IO.done(Exit.Failure(cause))
+          }
+      )
+
+  def context: ZIO[Telemetry, Nothing, SpanContext] =
+    getSpan.map(_.context)
+
+  def getBaggageItem(key: String): ZIO[Telemetry, Nothing, Option[String]] =
+    getSpan.map(_.getBaggageItem(key)).map(Option.apply)
+
+  def setBaggageItem(
+      key: String,
+      value: String
+  ): ZIO[Telemetry, Nothing, Unit] =
+    getSpan.flatMap(span => UIO(span.setBaggageItem(key, value))).unit
+
+  def tag(key: String, value: String): ZIO[Telemetry, Nothing, Unit] =
+    getSpan.flatMap(span => UIO(span.setTag(key, value))).unit
+
+  def tag(key: String, value: Int): ZIO[Telemetry, Nothing, Unit] =
+    getSpan.flatMap(span => UIO(span.setTag(key, value))).unit
+
+  def tag(key: String, value: Boolean): ZIO[Telemetry, Nothing, Unit] =
+    getSpan.flatMap(span => UIO(span.setTag(key, value))).unit
+
+  def tag[T <: Object](key: Tag[T], value: T): ZIO[Telemetry, Nothing, Unit] =
+    getSpan.flatMap(span => UIO(span.setTag(key, value))).unit
+
+  def log(msg: String): ZIO[COT, Nothing, Unit] =
+    for {
+      span <- getSpan
+      now <- getCurrentTimeMicros
+      _ <- UIO(span.log(now, msg))
+    } yield ()
+
+  def log(fields: Map[String, _]): ZIO[COT, Nothing, Unit] =
+    for {
+      span <- getSpan
+      now <- getCurrentTimeMicros
+      _ <- UIO(span.log(now, fields.asJava))
+    } yield ()
+
+  private def getSpan: ZIO[Telemetry, Nothing, Span] =
+    ZIO.accessM[Telemetry](_.telemetry.spanRef.get)
+
+  private def setSpan(span: Span): ZIO[Telemetry, Nothing, Unit] =
+    ZIO.accessM[Telemetry](_.telemetry.spanRef.set(span))
+
+  private def getTracer: ZIO[Telemetry, Nothing, Tracer] =
+    ZIO.environment[Telemetry].map(_.telemetry.tracer)
+
+  private def getCurrentTimeMicros: ZIO[Clock, Nothing, Long] =
+    ZIO.accessM[Clock](_.clock.currentTime(TimeUnit.MICROSECONDS))
+
+  implicit class Implicits[R, E, A](zio: ZIO[R, E, A]) extends AnyRef {
```

Comment:
I'd move this one to package object and make it extend `AnyVal`. Perhaps, I'd rename it to `TelemetrySyntax` just to align it with somehow standardized naming.